Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp

```diff
@@ -0,0 +1,576 @@
+    bool isNull() const { return m_isNull; }
+
+#if OS(ANDROID)
+    enum class Type : uint8_t {
+        Unset = 0,
+        FileDescriptor,
+        HardwareBuffer,
+    };
+
+    Type type() const { return m_type; }
+    void setType(Type type) { m_type = type; }
+#endif // OS(ANDROID)
+
+private:
+    // The AttachmentInfo will be copied using memcpy, so all members must be trivially copyable.
+    bool m_isNull;
+#if OS(ANDROID)
+    Type m_type;
+#endif
+
+};
+
+static_assert(sizeof(MessageInfo) + sizeof(AttachmentInfo) * s_attachmentMaxAmount <= s_messageMaxSize, "s_messageMaxSize is too small.");
+
+void Connection::platformInitialize(Identifier&& identifier)
+{
+    GUniqueOutPtr<GError> error;
+    m_socket = adoptGRef(g_socket_new_from_fd(identifier.handle.release(), &error.outPtr()));
+    if (!m_socket) {
+        // Note: g_socket_new_from_fd() takes ownership of the fd only on success, so if this error
+        // were not fatal, we would need to close it here.
+        g_error("Failed to adopt IPC::Connection socket: %s", error->message);
+    }
+    g_socket_set_blocking(m_socket.get(), FALSE);
+
+    m_cancellable = adoptGRef(g_cancellable_new());
+    m_readBuffer.reserveInitialCapacity(s_messageMaxSize);
+    m_fileDescriptors.reserveInitialCapacity(s_attachmentMaxAmount);
+}
+
+void Connection::platformInvalidate()
+{
+    GUniqueOutPtr<GError> error;
+    g_socket_close(m_socket.get(), &error.outPtr());
+    if (error)
+        RELEASE_LOG_ERROR(IPC, "Failed to close WebKit IPC socket: %s", error->message);
+
+    if (!m_isConnected)
+        return;
+
+    g_cancellable_cancel(m_cancellable.get());
+    m_readSocketMonitor.stop();
+    m_writeSocketMonitor.stop();
+
+    m_isConnected = false;
+}
+
+std::unique_ptr<Decoder> Connection::createMessageDecoder()
+{
+    if (m_readBuffer.size() < sizeof(MessageInfo)) {
+        RELEASE_LOG_FAULT(IPC, "createMessageDecoder: read buffer size is smaller than MessageInfo");
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    auto messageData = m_readBuffer.mutableSpan();
+    auto& messageInfo = consumeAndReinterpretCastTo<MessageInfo>(messageData);
+    if (messageInfo.attachmentCount() > s_attachmentMaxAmount || (!messageInfo.isBodyOutOfLine() && messageInfo.bodySize() > s_messageMaxSize)) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    auto attachmentCount = messageInfo.attachmentCount();
+    if (!attachmentCount)
+        return Decoder::create(messageData.first(messageInfo.bodySize()), { });
+
+    if (messageInfo.isBodyOutOfLine())
+        attachmentCount--;
+
+    Vector<Attachment> attachments(attachmentCount);
+    size_t fdIndex = 0;
+    for (size_t i = 0; i < attachmentCount; ++i) {
+        auto& attachmentInfo = consumeAndReinterpretCastTo<AttachmentInfo>(messageData);
+        size_t attachmentIndex = attachmentCount - i - 1;
+#if OS(ANDROID)
+        switch (attachmentInfo[i].type()) {
```

Comment:
It looks like you carried over the indexing from `ConnectionUnix.cpp`:

```suggestion
        switch (attachmentInfo.type()) {
```